From 99efe5bfd80982e7aa03f233fd1e871183229dba Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Mat=C3=ADas=20Hern=C3=A1ndez?= <matiashe@google.com>
Date: Fri, 4 Jul 2025 18:18:06 +0200
Subject: [PATCH 04/56] Cap the number of channels that an NLS can create for
 other packages

Also tweaked createConversationNotificationChannelForPackage to fix several edge conditions that could result in NPEs.

Bug: 425662627
Test: atest NotificationManagerServiceTest + manual
Flag: EXEMPT Bugfix
(cherry picked from commit 16b6836078042fc1b2288912e9656d01d9d20af5)
Cherrypick-From: https://googleplex-android-review.googlesource.com/q/commit:262f537110110e85eb0aaf27fbf6ee38a1659b05
Merged-In: I21033ce09185992adc0a377c701a442e8e81a9be
Change-Id: I21033ce09185992adc0a377c701a442e8e81a9be
---
 .../NotificationListenerStats.java            | 134 +++++++++++
 .../NotificationManagerService.java           |  79 +++++--
 .../NotificationManagerServiceTest.java       | 211 +++++++++---------
 .../server/notification/RoleObserverTest.java |   3 +-
 4 files changed, 302 insertions(+), 125 deletions(-)
 create mode 100644 services/core/java/com/android/server/notification/NotificationListenerStats.java

diff --git a/services/core/java/com/android/server/notification/NotificationListenerStats.java b/services/core/java/com/android/server/notification/NotificationListenerStats.java
new file mode 100644
index 000000000..3584e3176
--- /dev/null
+++ b/services/core/java/com/android/server/notification/NotificationListenerStats.java
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2025 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.android.server.notification;
+
+import android.util.SparseIntArray;
+
+import com.android.internal.annotations.GuardedBy;
+import com.android.internal.annotations.VisibleForTesting;
+import com.android.modules.utils.TypedXmlPullParser;
+import com.android.modules.utils.TypedXmlSerializer;
+import com.android.server.notification.ManagedServices.ManagedServiceInfo;
+
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+
+import java.io.IOException;
+import java.io.PrintWriter;
+
+class NotificationListenerStats {
+    /**
+     * Maximum number of channels that a single NLS is allowed to create for <em>all</em> installed
+     * packages. Should be much higher than reasonably necessary, but still prevent runaway channel
+     * creation abuse.
+     *
+     * @see PreferencesHelper#NOTIFICATION_CHANNEL_COUNT_LIMIT
+     */
+    private static final int MAX_CHANNELS = 5_000;
+
+    private static final String TAG_STATS = "nlsStats";
+    private static final String TAG_NLS = "nls";
+    private static final String ATT_UID = "uid";
+    private static final String ATT_OWNED_CHANNEL_COUNT = "channelCount";
+
+    private final Object mLock = new Object();
+
+    // nlsUid => count of channels created by that (privileged) NLS.
+    @GuardedBy("mLock")
+    private final SparseIntArray mChannelsCreated = new SparseIntArray();
+
+    private final int mMaxChannelsAllowed;
+
+    NotificationListenerStats() {
+        this(MAX_CHANNELS);
+    }
+
+    @VisibleForTesting
+    NotificationListenerStats(int maxChannelsAllowed) {
+        mMaxChannelsAllowed = maxChannelsAllowed;
+    }
+
+    boolean isAllowedToCreateChannel(ManagedServiceInfo nls) {
+        synchronized (mLock) {
+            int numCreated = mChannelsCreated.get(nls.uid);
+            return numCreated < mMaxChannelsAllowed;
+        }
+    }
+
+    void logCreatedChannels(ManagedServiceInfo nls, int increase) {
+        synchronized (mLock) {
+            int prevCreated = mChannelsCreated.get(nls.uid);
+            mChannelsCreated.put(nls.uid, prevCreated + increase);
+        }
+    }
+
+    void onPackageRemoved(int uid, String packageName) {
+        // If the uninstalled package was an NLS, drop its stats.
+        synchronized (mLock) {
+            mChannelsCreated.delete(uid);
+        }
+    }
+
+    static boolean isXmlTag(String tag) {
+        return TAG_STATS.equals(tag);
+    }
+
+    void writeXml(TypedXmlSerializer out) throws IOException {
+        out.startTag(null, TAG_STATS);
+        synchronized (mLock) {
+            for (int i = 0; i < mChannelsCreated.size(); i++) {
+                out.startTag(null, TAG_NLS);
+                out.attributeInt(null, ATT_UID, mChannelsCreated.keyAt(i));
+                out.attributeInt(null, ATT_OWNED_CHANNEL_COUNT, mChannelsCreated.valueAt(i));
+                out.endTag(null, TAG_NLS);
+            }
+        }
+        out.endTag(null, TAG_STATS);
+    }
+
+    void readXml(TypedXmlPullParser parser) throws XmlPullParserException, IOException {
+        int type = parser.getEventType();
+        if (type != XmlPullParser.START_TAG) return;
+        String tag = parser.getName();
+        if (!TAG_STATS.equals(tag)) return;
+
+        synchronized (mLock) {
+            while ((type = parser.next()) != XmlPullParser.END_DOCUMENT) {
+                tag = parser.getName();
+                if (type == XmlPullParser.END_TAG && TAG_STATS.equals(tag)) {
+                    break;
+                }
+                if (type == XmlPullParser.START_TAG) {
+                    if (TAG_NLS.equals(tag)) {
+                        int uid = parser.getAttributeInt(null, ATT_UID, 0);
+                        int channelCount = parser.getAttributeInt(null, ATT_OWNED_CHANNEL_COUNT, 0);
+                        mChannelsCreated.put(uid, channelCount);
+                    }
+                }
+            }
+        }
+    }
+
+    void dump(PrintWriter pw, String prefix) {
+        synchronized (mLock) {
+            for (int i = 0; i < mChannelsCreated.size(); i++) {
+                pw.println(prefix + "NLS with uid " + mChannelsCreated.keyAt(i));
+                pw.println(prefix + "  created channel count: " + mChannelsCreated.valueAt(i));
+            }
+        }
+    }
+}
diff --git a/services/core/java/com/android/server/notification/NotificationManagerService.java b/services/core/java/com/android/server/notification/NotificationManagerService.java
index a40afda0e..b55fc31dd 100644
--- a/services/core/java/com/android/server/notification/NotificationManagerService.java
+++ b/services/core/java/com/android/server/notification/NotificationManagerService.java
@@ -777,6 +777,7 @@ public class NotificationManagerService extends SystemService {
     @VisibleForTesting
     NotificationAssistants mAssistants;
     private ConditionProviders mConditionProviders;
+    private NotificationListenerStats mNotificationListenerStats;
     private NotificationUsageStats mUsageStats;
     private boolean mLockScreenAllowSecureNotifications = true;
     final ArrayMap<String, ArrayMap<Integer,
@@ -1229,6 +1230,9 @@ public class NotificationManagerService extends SystemService {
                 mLockScreenAllowSecureNotifications = parser.getAttributeBoolean(null,
                         LOCKSCREEN_ALLOW_SECURE_NOTIFICATIONS_VALUE, true);
             }
+            if (NotificationListenerStats.isXmlTag(parser.getName())) {
+                mNotificationListenerStats.readXml(parser);
+            }
         }
 
         if (!migratedManagedServices) {
@@ -1349,6 +1353,9 @@ public class NotificationManagerService extends SystemService {
         if (!forBackup || userId == USER_SYSTEM) {
             writeSecureNotificationsPolicy(out);
         }
+        if (!forBackup) {
+            mNotificationListenerStats.writeXml(out);
+        }
         out.endTag(null, TAG_NOTIFICATION_POLICY);
         out.endDocument();
     }
@@ -2722,7 +2729,8 @@ public class NotificationManagerService extends SystemService {
             SystemUiSystemPropertiesFlags.FlagResolver flagResolver,
             PermissionManager permissionManager, PowerManager powerManager,
             PostNotificationTrackerFactory postNotificationTrackerFactory,
-            UiEventLogger uiEventLogger, BitmapOffloadInternal bitmapOffloader) {
+            UiEventLogger uiEventLogger, BitmapOffloadInternal bitmapOffloader,
+            NotificationListenerStats notificationListenerStats) {
         mHandler = handler;
         if (Flags.nmBinderPerfThrottleEffectsSuppressorBroadcast()) {
             mBroadcastsHandler = broadcastsHandler;
@@ -2773,6 +2781,7 @@ public class NotificationManagerService extends SystemService {
         mMetricsLogger = new MetricsLogger();
         mRankingHandler = rankingHandler;
         mConditionProviders = conditionProviders;
+        mNotificationListenerStats = notificationListenerStats;
         mZenModeHelper = new ZenModeHelper(getContext(), mHandler.getLooper(), Clock.systemUTC(),
                 mConditionProviders, flagResolver, new ZenModeEventLogger(mPackageManagerClient));
         mZenModeHelper.addCallback(new ZenModeHelper.Callback() {
@@ -3096,7 +3105,7 @@ public class NotificationManagerService extends SystemService {
                 getContext().getSystemService(PermissionManager.class),
                 getContext().getSystemService(PowerManager.class),
                 new PostNotificationTrackerFactory() {}, new UiEventLoggerImpl(),
-                bitmapOffloader);
+                bitmapOffloader, new NotificationListenerStats());
 
         publishBinderService(Context.NOTIFICATION_SERVICE, mService, /* allowIsolated= */ false,
                 DUMP_FLAG_PRIORITY_CRITICAL | DUMP_FLAG_PRIORITY_NORMAL);
@@ -7245,7 +7254,15 @@ public class NotificationManagerService extends SystemService {
             Objects.requireNonNull(user);
             Objects.requireNonNull(parentId);
             Objects.requireNonNull(conversationId);
-            verifyPrivilegedListener(token, user, true);
+
+            ManagedServiceInfo nlsInfo = verifyPrivilegedListener(token, user, true);
+            if (!nlsInfo.isSystemUi()
+                    && !isNotificationAssistant(nlsInfo.service)
+                    && !mNotificationListenerStats.isAllowedToCreateChannel(nlsInfo)) {
+                Slog.e(TAG, "NLS " + nlsInfo + " has created too many channels already! "
+                        + "Rejecting " + pkg + "/" + user + "/" + parentId + "/" + conversationId);
+                return null;
+            }
 
             if (notificationClassification()) {
                 if (SYSTEM_RESERVED_IDS.contains(parentId)) {
@@ -7256,18 +7273,39 @@ public class NotificationManagerService extends SystemService {
                 }
             }
             int uid = getUidForPackageAndUser(pkg, user);
-            NotificationChannel conversationChannel =
-                    mPreferencesHelper.getNotificationChannel(pkg, uid, parentId, false).copy();
-            String conversationChannelId = String.format(
-                    CONVERSATION_CHANNEL_ID_FORMAT, parentId, conversationId);
+            if (uid == INVALID_UID) {
+                return null;
+            }
+            NotificationChannel parentChannel =
+                    mPreferencesHelper.getNotificationChannel(pkg, uid, parentId, false);
+            if (parentChannel == null) {
+                return null;
+            }
+
+            NotificationChannel previous = mPreferencesHelper.getConversationNotificationChannel(
+                    pkg, uid, parentId, conversationId, false, false);
+            if (previous != null) {
+                // If the conversation already exists, we're done. Continuing is worse since
+                // it would override any conversation customizations with the parent's values.
+                return previous;
+            }
+
+            String conversationChannelId = String.format(CONVERSATION_CHANNEL_ID_FORMAT, parentId,
+                    conversationId);
+            NotificationChannel conversationChannel = parentChannel.copy();
             conversationChannel.setId(conversationChannelId);
             conversationChannel.setConversationId(parentId, conversationId);
             createNotificationChannelsImpl(
-                    pkg, uid, new ParceledListSlice(Arrays.asList(conversationChannel)));
-            handleSavePolicyFile();
+                    pkg, uid, new ParceledListSlice<>(Arrays.asList(conversationChannel)));
 
-            return mPreferencesHelper.getConversationNotificationChannel(
-                    pkg, uid, parentId, conversationId, false, false).copy();
+            NotificationChannel created = mPreferencesHelper.getConversationNotificationChannel(
+                    pkg, uid, parentId, conversationId, false, false);
+            if (created != null) {
+                mNotificationListenerStats.logCreatedChannels(nlsInfo, /* increase= */ 1);
+                handleSavePolicyFile();
+            }
+
+            return created;
         }
 
         @Override
@@ -7359,8 +7397,9 @@ public class NotificationManagerService extends SystemService {
             return mPermissionHelper.isPermissionFixed(pkg, userId);
         }
 
-        private void verifyPrivilegedListener(INotificationListener token, UserHandle user,
-                boolean assistantAllowed) {
+        @NonNull
+        private ManagedServiceInfo verifyPrivilegedListener(INotificationListener token,
+                UserHandle user, boolean assistantAllowed) {
             ManagedServiceInfo info;
             synchronized (mNotificationLock) {
                 info = mListeners.checkServiceTokenLocked(token);
@@ -7375,6 +7414,7 @@ public class NotificationManagerService extends SystemService {
             if (!info.enabledAndUserMatches(user.getIdentifier())) {
                 throw new SecurityException(info + " does not have access");
             }
+            return info;
         }
 
         private void verifyPrivilegedListenerUriPermission(int sourceUid,
@@ -8084,6 +8124,7 @@ public class NotificationManagerService extends SystemService {
                 pw.println("\n  Notification listeners:");
                 mListeners.dump(pw, filter);
                 pw.print("    mListenerHints: "); pw.println(mListenerHints);
+
                 pw.print("    mListenersDisablingEffects: (");
                 N = mListenersDisablingEffects.size();
                 for (int i = 0; i < N; i++) {
@@ -8103,6 +8144,10 @@ public class NotificationManagerService extends SystemService {
                     }
                 }
                 pw.println(')');
+
+                pw.println("\n  NotificationListenerStats:");
+                mNotificationListenerStats.dump(pw, "    ");
+
                 pw.println("\n  Notification assistant services:");
                 mAssistants.dump(pw, filter);
             }
@@ -10899,6 +10944,8 @@ public class NotificationManagerService extends SystemService {
                 // (recently dismissed notifications) and notification history.
                 mArchive.removePackageNotifications(pkg, userHandle);
                 mHistoryManager.onPackageRemoved(userHandle, pkg);
+                // Remove from NLS Stats (in case the package included an NLS).
+                mNotificationListenerStats.onPackageRemoved(uid, pkg);
             }
         }
         if (preferencesChanged) {
@@ -12214,11 +12261,11 @@ public class NotificationManagerService extends SystemService {
      */
     @VisibleForTesting
     boolean isInteractionVisibleToListener(ManagedServiceInfo info, int userId) {
-        boolean isAssistantService = isServiceTokenValid(info.getService());
+        boolean isAssistantService = isNotificationAssistant(info.getService());
         return !isAssistantService || info.isSameUser(userId);
     }
 
-    private boolean isServiceTokenValid(IInterface service) {
+    private boolean isNotificationAssistant(IInterface service) {
         synchronized (mNotificationLock) {
             return mAssistants.isServiceTokenValidLocked(service);
         }
@@ -14463,7 +14510,7 @@ public class NotificationManagerService extends SystemService {
                 BackgroundThread.getHandler().post(() -> {
                     if (info.isSystem
                             || hasCompanionDevice(info)
-                            || isServiceTokenValid(info.service)) {
+                            || isNotificationAssistant(info.service)) {
                         notifyNotificationChannelChanged(
                                 info, pkg, user, channel, modificationType);
                     }
diff --git a/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java b/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java
index 152d625af..3e0af225a 100644
--- a/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java
+++ b/services/tests/uiservicestests/src/com/android/server/notification/NotificationManagerServiceTest.java
@@ -425,6 +425,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                     .setOwner(new ComponentName("pkg", "cls"))
                     .build();
 
+    private static final int MAX_CHANNELS_CREATED_BY_NLS_FOR_TESTING = 10;
+
     @ClassRule
     public static final LimitDevicesRule sLimitDevicesRule = new LimitDevicesRule();
 
@@ -512,16 +514,6 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
 
     NotificationChannel mMinChannel = new NotificationChannel("min", "min", IMPORTANCE_MIN);
 
-    private final NotificationChannel mParentChannel =
-            new NotificationChannel(PARENT_CHANNEL_ID, "parentName", IMPORTANCE_DEFAULT);
-    private final NotificationChannel mConversationChannel =
-            new NotificationChannel(
-                    CONVERSATION_CHANNEL_ID, "conversationName", IMPORTANCE_DEFAULT);
-
-    private static final String PARENT_CHANNEL_ID = "parentChannelId";
-    private static final String CONVERSATION_CHANNEL_ID = "conversationChannelId";
-    private static final String CONVERSATION_ID = "conversationId";
-
     private static final int NOTIFICATION_LOCATION_UNKNOWN = 0;
 
     private static final String VALID_CONVO_SHORTCUT_ID = "shortcut";
@@ -685,6 +677,7 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         // MockPackageManager - default returns ApplicationInfo with matching calling UID
         mContext.setMockPackageManager(mPackageManagerClient);
 
+        when(mPackageManager.getPackageUid(eq(mPkg), anyLong(), eq(mUserId))).thenReturn(mUid);
         when(mPackageManager.getApplicationInfo(anyString(), anyLong(), anyInt()))
                 .thenAnswer((Answer<ApplicationInfo>) invocation -> {
                     Object[] args = invocation.getArguments();
@@ -834,7 +827,8 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
                 mAppOpsManager, mUm, mHistoryManager, mStatsManager, mAmi, mToastRateLimiter,
                 mPermissionHelper, mock(UsageStatsManagerInternal.class), mTelecomManager, mLogger,
                 mTestFlagResolver, mPermissionManager, mPowerManager,
-                mPostNotificationTrackerFactory, mUiEventLogger, mBitmapOffloader);
+                mPostNotificationTrackerFactory, mUiEventLogger, mBitmapOffloader,
+                new NotificationListenerStats(MAX_CHANNELS_CREATED_BY_NLS_FOR_TESTING));
 
         mService.setAttentionHelper(mAttentionHelper);
         mService.setLockPatternUtils(mock(LockPatternUtils.class));
@@ -4937,46 +4931,22 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         verify(mAmi).hasForegroundServiceNotification(anyString(), anyInt(), anyString());
     }
 
-    private void setUpChannelsForConversationChannelTest() throws RemoteException {
-        when(mPreferencesHelper.getNotificationChannel(
-                eq(mPkg), eq(mUid), eq(PARENT_CHANNEL_ID), eq(false)))
-                .thenReturn(mParentChannel);
-        when(mPreferencesHelper.getConversationNotificationChannel(
-                eq(mPkg), eq(mUid), eq(PARENT_CHANNEL_ID), eq(CONVERSATION_ID), eq(false), eq(false)))
-                .thenReturn(mConversationChannel);
-        when(mPackageManager.getPackageUid(mPkg, 0, mUserId)).thenReturn(mUid);
-    }
-
     @Test
     @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
     public void createConversationChannelForPkgFromPrivilegedListener_cdm_success() throws Exception {
-        // Set up cdm
-        mService.setPreferencesHelper(mPreferencesHelper);
         when(mCompanionMgr.getAssociations(mPkg, mUserId))
                 .thenReturn(singletonList(mock(AssociationInfo.class)));
+        mService.mPreferencesHelper.createNotificationChannel(mPkg, mUid,
+                new NotificationChannel("parent", "parentName", IMPORTANCE_DEFAULT),
+                true, false, mUid, false);
 
-        // Set up parent channel
-        setUpChannelsForConversationChannelTest();
-        final NotificationChannel parentChannelCopy = mParentChannel.copy();
+        NotificationChannel createdChannel = mBinderService
+                .createConversationNotificationChannelForPackageFromPrivilegedListener(
+                    null, mPkg, mUser, "parent", "convo");
 
-        NotificationChannel createdChannel =
-                mBinderService.createConversationNotificationChannelForPackageFromPrivilegedListener(
-                    null, mPkg, mUser, PARENT_CHANNEL_ID, CONVERSATION_ID);
-
-        // Verify that a channel is created and a copied channel is returned.
-        verify(mPreferencesHelper, times(1)).createNotificationChannel(
-                eq(mPkg), eq(mUid), any(), anyBoolean(), anyBoolean(),
-                eq(mUid), anyBoolean());
-        assertThat(createdChannel).isNotSameInstanceAs(mConversationChannel);
-        assertThat(createdChannel).isEqualTo(mConversationChannel);
-
-        // Verify that the channel creation is not directly use the parent channel.
-        verify(mPreferencesHelper, never()).createNotificationChannel(
-                anyString(), anyInt(), eq(mParentChannel), anyBoolean(), anyBoolean(),
-                anyInt(), anyBoolean());
-
-        // Verify that the content of parent channel is not changed.
-        assertThat(parentChannelCopy).isEqualTo(mParentChannel);
+        assertThat(createdChannel).isNotNull();
+        assertThat(createdChannel.getParentChannelId()).isEqualTo("parent");
+        assertThat(createdChannel.getConversationId()).isEqualTo("convo");
     }
 
     @Test
@@ -5012,89 +4982,56 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
     @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
     public void createConversationChannelForPkgFromPrivilegedListener_cdm_noAccess() throws Exception {
         // Set up cdm without access
-        mService.setPreferencesHelper(mPreferencesHelper);
-        when(mCompanionMgr.getAssociations(mPkg, mUserId))
-                .thenReturn(emptyList());
+        when(mCompanionMgr.getAssociations(mPkg, mUserId)).thenReturn(emptyList());
+        mService.mPreferencesHelper.createNotificationChannel(mPkg, mUid,
+                new NotificationChannel("parent", "parentName", IMPORTANCE_DEFAULT),
+                true, false, mUid, false);
 
-        // Set up parent channel
-        setUpChannelsForConversationChannelTest();
-
-        try {
-            mBinderService.createConversationNotificationChannelForPackageFromPrivilegedListener(
-                null, mPkg, mUser, "parentId", "conversationId");
-            fail("listeners that don't have a companion device shouldn't be able to call this");
-        } catch (SecurityException e) {
-            // pass
-        }
-
-        verify(mPreferencesHelper, never()).createNotificationChannel(
-                anyString(), anyInt(), any(), anyBoolean(), anyBoolean(),
-                anyInt(), anyBoolean());
+        assertThrows(SecurityException.class, () ->
+                mBinderService
+                        .createConversationNotificationChannelForPackageFromPrivilegedListener(null,
+                                mPkg, mUser, "parent", "convo"));
     }
 
     @Test
     @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
     public void createConversationChannelForPkgFromPrivilegedListener_assistant_success() throws Exception {
         // Set up assistant
-        mService.setPreferencesHelper(mPreferencesHelper);
         when(mCompanionMgr.getAssociations(mPkg, mUserId))
                 .thenReturn(emptyList());
         when(mAssistants.isServiceTokenValidLocked(any())).thenReturn(true);
-
-        // Set up parent channel
-        setUpChannelsForConversationChannelTest();
-        final NotificationChannel parentChannelCopy = mParentChannel.copy();
+        mService.mPreferencesHelper.createNotificationChannel(mPkg, mUid,
+                new NotificationChannel("parent", "parentName", IMPORTANCE_DEFAULT),
+                true, false, mUid, false);
 
         NotificationChannel createdChannel =
                 mBinderService.createConversationNotificationChannelForPackageFromPrivilegedListener(
-                    null, mPkg, mUser, PARENT_CHANNEL_ID, CONVERSATION_ID);
-
-        // Verify that a channel is created and a copied channel is returned.
-        verify(mPreferencesHelper, times(1)).createNotificationChannel(
-                eq(mPkg), eq(mUid), any(), anyBoolean(), anyBoolean(),
-                eq(mUid), anyBoolean());
-        assertThat(createdChannel).isNotSameInstanceAs(mConversationChannel);
-        assertThat(createdChannel).isEqualTo(mConversationChannel);
-
-        // Verify that the channel creation is not directly use the parent channel.
-        verify(mPreferencesHelper, never()).createNotificationChannel(
-                anyString(), anyInt(), eq(mParentChannel), anyBoolean(), anyBoolean(),
-                anyInt(), anyBoolean());
+                    null, mPkg, mUser, "parent", "convo");
 
-        // Verify that the content of parent channel is not changed.
-        assertThat(parentChannelCopy).isEqualTo(mParentChannel);
+        assertThat(createdChannel).isNotNull();
     }
 
     @Test
     @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
     public void createConversationChannelForPkgFromPrivilegedListener_assistant_noAccess() throws Exception {
         // Set up assistant without access
-        mService.setPreferencesHelper(mPreferencesHelper);
         when(mCompanionMgr.getAssociations(mPkg, mUserId))
                 .thenReturn(emptyList());
         when(mAssistants.isServiceTokenValidLocked(any())).thenReturn(false);
+        mService.mPreferencesHelper.createNotificationChannel(mPkg, mUid,
+                new NotificationChannel("parent", "parentName", IMPORTANCE_DEFAULT),
+                true, false, mUid, false);
 
-        // Set up parent channel
-        setUpChannelsForConversationChannelTest();
-
-        try {
-            mBinderService.createConversationNotificationChannelForPackageFromPrivilegedListener(
-                null, mPkg, mUser, "parentId", "conversationId");
-            fail("listeners that don't have a companion device shouldn't be able to call this");
-        } catch (SecurityException e) {
-            // pass
-        }
-
-        verify(mPreferencesHelper, never()).createNotificationChannel(
-                anyString(), anyInt(), any(), anyBoolean(), anyBoolean(),
-                anyInt(), anyBoolean());
+        assertThrows(SecurityException.class, () ->
+                mBinderService
+                        .createConversationNotificationChannelForPackageFromPrivilegedListener(null,
+                                mPkg, mUser, "parent", "convo"));
     }
 
     @Test
     @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
     public void createConversationChannelForPkgFromPrivilegedListener_badUser() throws Exception {
         // Set up bad user
-        mService.setPreferencesHelper(mPreferencesHelper);
         when(mCompanionMgr.getAssociations(mPkg, mUserId))
                 .thenReturn(singletonList(mock(AssociationInfo.class)));
         mListener = mock(ManagedServices.ManagedServiceInfo.class);
@@ -5102,20 +5039,78 @@ public class NotificationManagerServiceTest extends UiServiceTestCase {
         when(mListener.enabledAndUserMatches(anyInt())).thenReturn(false);
         when(mListeners.checkServiceTokenLocked(any())).thenReturn(mListener);
 
-        // Set up parent channel
-        setUpChannelsForConversationChannelTest();
+        assertThrows(SecurityException.class, () ->
+                mBinderService
+                        .createConversationNotificationChannelForPackageFromPrivilegedListener(null,
+                                mPkg, mUser, "parent", "convo"));
+    }
 
-        try {
-            mBinderService.createConversationNotificationChannelForPackageFromPrivilegedListener(
-                null, mPkg, mUser, "parentId", "conversationId");
-            fail("listener getting channels from a user they cannot see");
-        } catch (SecurityException e) {
-            // pass
+    @Test
+    @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
+    public void createConversationChannelForPkgFromPrivilegedListener_tooManyChannels_ignored()
+            throws Exception {
+        NotificationChannel parentChannel = mTestNotificationChannel.copy();
+        mService.mPreferencesHelper.createNotificationChannel(mPkg, mUid, parentChannel, true,
+                false, mUid, false);
+        when(mCompanionMgr.getAssociations(mPkg, mUserId))
+                .thenReturn(singletonList(mock(AssociationInfo.class)));
+
+        for (int i = 1; i <= MAX_CHANNELS_CREATED_BY_NLS_FOR_TESTING; i++) {
+            NotificationChannel createdConversation = mBinderService
+                    .createConversationNotificationChannelForPackageFromPrivilegedListener(
+                            null, mPkg, mUser, parentChannel.getId(), "conversation #" + i);
+            assertThat(createdConversation).isNotNull();
+            assertThat(createdConversation.getParentChannelId()).isEqualTo(parentChannel.getId());
         }
+        ParceledListSlice<NotificationChannel> packageChannels =
+                mService.mPreferencesHelper.getNotificationChannels(mPkg, mUid, false, false);
+        assertThat(packageChannels.getList().stream().filter(c -> c.isConversation()))
+                .hasSize(MAX_CHANNELS_CREATED_BY_NLS_FOR_TESTING);
+
+        // And finally, monsieur, a wafer-thin mint.
+        NotificationChannel oneTooMany = mBinderService
+                .createConversationNotificationChannelForPackageFromPrivilegedListener(
+                        null, mPkg, mUser, parentChannel.getId(), "another conversation");
+
+        assertThat(oneTooMany).isNull();
+        ParceledListSlice<NotificationChannel> unchangedPackageChannels =
+                mService.mPreferencesHelper.getNotificationChannels(mPkg, mUid, false, false);
+        assertThat(unchangedPackageChannels.getList()).isEqualTo(packageChannels.getList());
+    }
+
+    @Test
+    @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
+    public void createConversationChannelForPkgFromPrivilegedListener_noParentChannel_ignored()
+            throws Exception {
+        when(mCompanionMgr.getAssociations(mPkg, mUserId))
+                .thenReturn(singletonList(mock(AssociationInfo.class)));
+
+        NotificationChannel attempted = mBinderService
+                .createConversationNotificationChannelForPackageFromPrivilegedListener(
+                        null, mPkg, mUser, "missing_parent", "convo");
+        assertThat(attempted).isNull();
+    }
+
+    @Test
+    @RequiresFlagsEnabled(FLAG_NOTIFICATION_CONVERSATION_CHANNEL_MANAGEMENT)
+    public void createConversationChannelForPkgFromPrivilegedListener_existingChannel_returnsSame()
+            throws Exception {
+        NotificationChannel parentChannel = mTestNotificationChannel.copy();
+        mService.mPreferencesHelper.createNotificationChannel(mPkg, mUid, parentChannel, true,
+                false, mUid, false);
+        when(mCompanionMgr.getAssociations(mPkg, mUserId))
+                .thenReturn(singletonList(mock(AssociationInfo.class)));
+
+        NotificationChannel created = mBinderService
+                .createConversationNotificationChannelForPackageFromPrivilegedListener(
+                        null, mPkg, mUser, parentChannel.getId(), "convo");
+        assertThat(created).isNotNull();
+
+        NotificationChannel again =  mBinderService
+                .createConversationNotificationChannelForPackageFromPrivilegedListener(
+                        null, mPkg, mUser, parentChannel.getId(), "convo");
 
-        verify(mPreferencesHelper, never()).createNotificationChannel(
-                anyString(), anyInt(), any(), anyBoolean(), anyBoolean(),
-                anyInt(), anyBoolean());
+        assertThat(again).isSameInstanceAs(created);
     }
 
     @Test
diff --git a/services/tests/uiservicestests/src/com/android/server/notification/RoleObserverTest.java b/services/tests/uiservicestests/src/com/android/server/notification/RoleObserverTest.java
index fe2c3d8a2..5b6fd8aa7 100644
--- a/services/tests/uiservicestests/src/com/android/server/notification/RoleObserverTest.java
+++ b/services/tests/uiservicestests/src/com/android/server/notification/RoleObserverTest.java
@@ -175,7 +175,8 @@ public class RoleObserverTest extends UiServiceTestCase {
                     mock(PowerManager.class),
                     new NotificationManagerService.PostNotificationTrackerFactory() {},
                     mock(UiEventLogger.class),
-                    mock(BitmapOffloadInternal.class));
+                    mock(BitmapOffloadInternal.class),
+                    new NotificationListenerStats());
         } catch (SecurityException e) {
             if (!e.getMessage().contains("Permission Denial: not allowed to send broadcast")) {
                 throw e;
-- 
2.52.0

